// open-cloud-client/src/Viewer/Draggers/OdZoomDragger.js
///////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2002-2021, Open Design Alliance (the "Alliance").
// All rights reserved.
//
// This software and its documentation and related materials are owned by
// the Alliance. The software may only be incorporated into application
// programs owned by members of the Alliance, subject to a signed
// Membership Agreement and Supplemental Software License Agreement with the
// Alliance. The structure and organization of this software are the valuable
// trade secrets of the Alliance and its suppliers. The software is also
// protected by copyright law and international treaty provisions. Application
// programs incorporating this software must include the following statement
// with their copyright notices:
//
//   This application incorporates Open Design Alliance software pursuant to a
//   license agreement with Open Design Alliance.
//   Open Design Alliance Copyright (C) 2002-2021 by Open Design Alliance.
//   All rights reserved.
//
// By use of this software, its documentation or related materials, you
// acknowledge and accept the above terms.
///////////////////////////////////////////////////////////////////////////////
/* eslint-disable no-unused-vars */
import { OdBaseDragger } from './Common/OdBaseDragger.js';

export class OdZoomDragger extends OdBaseDragger {
  constructor(...args) {
    super(...args);
    this.press = false;

    this.getViewer().setEnableAutoSelect(true);
  }

  start(x, y) {
    this.press = true;
    this.pressX = x;
    this.pressY = y;
  }

  drag(x, y, dltX, dltY) {
    if (this.press) {

      const ZOOM_SPEED = 0.025
  
      const viewer = this.m_module.getViewer()

      viewer.zoomAt(dltY > 0 ? 1 + ZOOM_SPEED : 1 - ZOOM_SPEED, this.pressX, this.pressY)

      if (this.subject.activeDragger() && this.subject.activeDragger().updatePreview)
      {
        this.subject.activeDragger().updatePreview()
      }
    }
  }

  end(x, y) {
    this.press = false;
  }

  mouseleave(ev) {
    this.end();
  }
}
