// open-cloud-client/src/Api/Options.js
export class Options
{
  constructor(emitter)
  {
    this._showFPS = false
    this._showWCS = true
    this._cameraAnimation = true
    this._antialiasing = true
    this._groundShadow = false
    this._shadows = false
    this._cameraAxisXSpeed = 4;
    this._cameraAxisYSpeed = 1;
    this._ambientOcclusion = false
    this.loadFromStorage()
    this._emitter = emitter
  }

  notifierChangeEvent()
  {
    if (this._emitter)
    {
      this.saveToStorage()
      this._emitter.emit(
        {
          type: 'changeClientOption',
          data: this
        }
      )
    }
  }

  saveToStorage()
  {
    const opt = {
      showFPS: this._showFPS,
      showWCS: this._showWCS,
      cameraAnimation: this._cameraAnimation,
      antialiasing: this._antialiasing,
      groundShadow: this._groundShadow,
      shadows: this._shadows,
      cameraAxisXSpeed: this._cameraAxisXSpeed,
      cameraAxisYSpeed: this._cameraAxisYSpeed,
      ambientOcclusion: this._ambientOcclusion
    }
   
    localStorage.setItem('od-client-settings', JSON.stringify(opt))
  }

  loadFromStorage()
  {
    const jsonOpt = localStorage.getItem('od-client-settings')
    if (jsonOpt)
    {
      const opt = JSON.parse(jsonOpt)

      this._showFPS         = opt.showFPS
      this._showWCS         = opt.showWCS
      this._cameraAnimation = opt.cameraAnimation
      this._antialiasing    = opt.antialiasing
      this._groundShadow    = opt.groundShadow
      this._shadows         = opt.shadows
      this._cameraAxisXSpeed = opt.cameraAxisXSpeed || 4
      this._cameraAxisYSpeed = opt.cameraAxisYSpeed || 1
      this._ambientOcclusion = opt.ambientOcclusion || false
    }
  }

  get showFPS() {
    return this._showFPS
  }

  set showFPS(value) {
    this._showFPS = value
    this.notifierChangeEvent()
  }

  get showWCS() {
    return this._showWCS
  }

  set showWCS(value) {
    this._showWCS = value
    this.notifierChangeEvent()
  }

  get cameraAnimation() {
    return this._cameraAnimation
  }

  set cameraAnimation(value) {
    this._cameraAnimation = value
    this.notifierChangeEvent()
  }

  get antialiasing() {
    return this._antialiasing
  }

  set antialiasing(value) {
    this._antialiasing = value
    this.notifierChangeEvent()
  }

  get groundShadow() {
    return this._groundShadow
  }

  set groundShadow(value) {
    this._groundShadow = value
    this.notifierChangeEvent()
  }

  get shadows() {
    return this._shadows
  }

  set shadows(value) {
    this._shadows = value
    this.notifierChangeEvent()
  }
  
  set cameraAxisXSpeed(value) {
    this._cameraAxisXSpeed = value
    this.notifierChangeEvent()
  }

  get cameraAxisXSpeed() {
    return this._cameraAxisXSpeed
  }

  set cameraAxisYSpeed(value) {
    this._cameraAxisYSpeed = value
    this.notifierChangeEvent()
  }

  get cameraAxisYSpeed() {
    return this._cameraAxisYSpeed 
  }

  get ambientOcclusion() {
    return this._ambientOcclusion
  }

  set ambientOcclusion(value) {
    this._ambientOcclusion = value
    this.notifierChangeEvent()
  }
}
