// open-cloud-client/src/Api/Project.js
///////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2002-2021, Open Design Alliance (the "Alliance").
// All rights reserved.
//
// This software and its documentation and related materials are owned by
// the Alliance. The software may only be incorporated into application
// programs owned by members of the Alliance, subject to a signed
// Membership Agreement and Supplemental Software License Agreement with the
// Alliance. The structure and organization of this software are the valuable
// trade secrets of the Alliance and its suppliers. The software is also
// protected by copyright law and international treaty provisions. Application
// programs incorporating this software must include the following statement
// with their copyright notices:
//
//   This application incorporates Open Design Alliance software pursuant to a
//   license agreement with Open Design Alliance.
//   Open Design Alliance Copyright (C) 2002-2021 by Open Design Alliance.
//   All rights reserved.
//
// By use of this software, its documentation or related materials, you
// acknowledge and accept the above terms.
///////////////////////////////////////////////////////////////////////////////

import { ProjectImpl } from "./impl/ProjectImpl";

/**
 * @class
 */
 export class Project {
  constructor(impl, app) {
    this._impl = impl;
    this.app = app;
  }

  /**
   * Project id
   * @type {(string)}
   */
  get id() {
    return this._impl._data.id;
  }
  /**
   * Project name
   * @type {(string)}
   */
  get name() {
    return this._impl._data.name;
  }

  set name(value) {
    this._impl._data.name = value;
  }
  /**
   * Project authorization for current user
   * @type {(Object)}
   */
  get authorization() {
    return this._impl._data.authorization;
  }
  /**
   * Project created date in ISO string
   * @type {(string)}
   */
  get createdAt() {
    return this._impl._data.createdAt;
  }
  /**
   * Project last update date in ISO string
   * @type {(string)}
   */
  get updatedAt() {
    return this._impl._data.updatedAt;
  }
  /**
   * Project start date in ISO string
   * @type {(string)}
   */
  get startDate() {
    return this._impl._data.startDate;
  }

  set startDate(value) {
    this._impl._data.startDate = value.toISOString();
  }
  /**
   * Project start date in ISO string
   * @type {(string)}
   */
  get endDate() {
    return this._impl._data.endDate;
  }

  set endDate(value) {
    this._impl._data.endDate = value.toISOString();
  }
  /**
   * Project detail description
   * @type {(string)}
   */
  get description() {
    return this._impl._data.description;
  }

  set description(value) {
    this._impl._data.description = value;
  }
  /**
   * Project avatar url (base64 image)
   * @type {(string)}
   */
  get avatarUrl() {
    return this._impl._data.avatarUrl;
  }

  set avatarUrl(value) {
    this._impl._data.avatarUrl = value;
  }
  /**
   * Project custom fields
   * @type {(Object)}
   */
  get customFields() {
    return this._impl._data.customFields;
  }

  set customFields(value) {
    this._impl._data.customFields = value;
  }
  /**
   * Project public flag
   * @type {(Object)}
   */
  get public() {
    return this._impl._data.public;
  }

  set public(value) {
    this._impl._data.public = value;
  }

  /**
   * Save project
   */
  save() {
    return this._impl.save()
  }

  /**
   * Delete project
   */
  delete() {
    return this._impl.delete()
  }
}
